Sources/NIO/Selector.swift

```diff
@@ -614,48 +629,75 @@ final class Selector<R: Registration> {
     ///
     /// After closing the `Selector` it's no longer possible to use it.
     public func close() throws {
+        assert(self.myThread == NIOThread.current)
         guard self.lifecycleState == .open else {
             throw IOError(errnoCode: EBADF, reason: "can't close selector as it's \(self.lifecycleState).")
         }
         self.lifecycleState = .closed
 
         self.registrations.removeAll()
 
-        /* note, we can't close `self.fd` (on macOS) or `self.eventfd` (on Linux) here as that's read unprotectedly and might lead to race conditions. Instead, we abuse ARC to close it for us. */
+        /* note, we can't close `self.fd` (on macOS) or `self.eventFD` (on Linux) here as that's read unprotectedly and might lead to race conditions. Instead, we abuse ARC to close it for us. */
+        try self.externalSelectorFDLock.withLock {
 #if os(Linux)
-        _ = try Posix.close(descriptor: self.timerfd)
-#endif
+            defer {
+                // needs to be done in case we throw too.
+                self.eventFD = -1
+                self.timerFD = -1
+            }
 
-#if os(Linux)
-        /* `self.fd` is used as the event file descriptor to wake kevent() up so can't be closed here on macOS */
-        _ = try Posix.close(descriptor: self.fd)
+            _ = try Posix.close(descriptor: self.timerFD)
+            _ = try Posix.close(descriptor: self.eventFD)
 #endif
+            defer {
+                // needs to be done in case we throw too.
+                self.selectorFD = -1
+            }
+            try Posix.close(descriptor: self.selectorFD)
+        }
     }
 
     /* attention, this may (will!) be called from outside the event loop, ie. can't access mutable shared state (such as `self.open`) */
     func wakeup() throws {
-
-#if os(Linux)
-        /* this is fine as we're abusing ARC to close `self.eventfd` */
-        _ = try EventFd.eventfd_write(fd: self.eventfd, value: 1)
-#else
-        var event = kevent()
-        event.ident = 0
-        event.filter = Int16(EVFILT_USER)
-        event.fflags = UInt32(NOTE_TRIGGER | NOTE_FFNOP)
-        event.data = 0
-        event.udata = nil
-        event.flags = 0
-        try withUnsafeMutablePointer(to: &event) { ptr in
-            try kqueueApplyEventChangeSet(keventBuffer: UnsafeMutableBufferPointer(start: ptr, count: 1))
+        assert(NIOThread.current != self.myThread)
```

Comment:
some message for the assert or nah?

Reply:
@ktoso don't think it adds much value, we know it's the wrong thread and we get which one from the backtrace.

Reply:
Also true, your call :) I always make a point of doing including a message, a style thing more than anything I guess